packages/sdk/src/utils/transaction.ts

```diff
@@ -136,47 +108,119 @@ export async function sendTransactionSafely(
+          url: `${request.baseURL}${item?.check?.endpoint}`,
+          method: item?.check?.method,
+          headers: headers
+        })
+      }
+      if (!res || validate(res)) {
+        waitingForConfirmation = false // transaction confirmed
+      } else if (res) {
+        if (res.data.status !== 'pending') {
+          attemptCount++
+        }
+
+        await new Promise((resolve) => setTimeout(resolve, pollingInterval))
       }
+    }
+
+    if (attemptCount >= maximumAttempts) {
+      throw new SolverStatusTimeoutError(txHash as Address, attemptCount)
+    }
 
-      await new Promise((resolve) => setTimeout(resolve, pollingInterval))
+    if (transactionCancelled) {
+      throw Error('Transaction was cancelled')
     }
+    return true
   }
 
-  if (attemptCount >= maximumAttempts) {
-    throw new SolverStatusTimeoutError(txHash, attemptCount)
+  const waitForTransaction = () => {
+    const controller = new AbortController()
+    const signal = controller.signal
+    // Handle transaction replacements and cancellations
+    return {
+      promise: viemClient
+        .waitForTransactionReceipt({
+          hash: txHash as Address,
+          onReplaced: (replacement) => {
+            if (signal.aborted) {
+              return
+            }
+            if (replacement.reason === 'cancelled') {
+              transactionCancelled = true
+              throw Error('Transaction cancelled')
+            }
+
+            setTxHashes([
+              { txHash: replacement.transaction.hash, chainId: chainId }
+            ])
+            txHash = replacement.transaction.hash
+            attemptCount = 0 // reset attempt count
+            getClient()?.log(
+              ['Transaction replaced', replacement],
+              LogLevel.Verbose
+            )
+            postTransactionToSolver({
+              txHash,
+              chainId,
+              step,
+              request,
+              headers
+            })
+          }
+        })
+        .then((data) => {
+          if (signal.aborted) {
+            return
+          }
+          receipt = data
+          getClient()?.log(
+            ['Transaction Receipt obtained', receipt],
+            LogLevel.Verbose
+          )
+        })
+        .catch((error) => {
+          if (signal.aborted) {
+            return
+          }
+          getClient()?.log(
+            ['Error in waitForTransactionReceipt', error],
+            LogLevel.Error
+          )
+        }),
+      controller
+    }
   }
 
-  if (transactionCancelled) {
-    throw Error('Transaction was cancelled')
+  //Sequence internal functions
+  if (step.id === 'approve') {
+    await waitForTransaction().promise
+    await pollForConfirmation()
+  } else {
+    const { promise: receiptPromise, controller: receiptController } =
+      waitForTransaction()
+    const confirmationPromise = pollForConfirmation()
+
+    await Promise.race([receiptPromise, confirmationPromise])
```

Comment:
Do we even need to race these promises if we are always going to require the confirmation promise?

Reply:
It could be that the receipt finishes first.